packages/deployments/contracts/contracts/nomad-xapps/contracts/promise-router/PromiseRouter.sol

```diff
@@ -154,23 +211,58 @@ contract PromiseRouter is Version0, Router {
 
     bytes32 transferId = _msg.transferId();
     address callbackAddress = _msg.callbackAddress();
-    bytes memory data = _msg.returnCallData();
+    bool success = _msg.returnSuccess();
+    bytes memory data = _msg.returnData();
 
-    //TODO process callback
+    //store Promise message
+    promiseMessages[transferId] = _message;
 
     // emit Receive event
-    emit Receive(_originAndNonce(_origin, _nonce), _origin, transferId, callbackAddress, data);
+    emit Receive(_originAndNonce(_origin, _nonce), _origin, transferId, callbackAddress, success, data, _message);
   }
 
-  function process(bytes32 transactionId, bytes memory _message) internal {
-    // Should parse out the return data and callback address from message
-    // Should execute the callback() function on the provided Callback address
-    // Should enforce relayer is whitelisted by calling local connext contract
+  function process(bytes32 transferId) public payable nonReentrant {
+    // parse out the return data and callback address from message
+    bytes memory _message = promiseMessages[transferId];
+    if (_message.length == 0) revert PromiseRouter__process_InvalidMessage();
+
+    bytes29 _msg = _message.ref(0).mustBePromiseCallback();
+
+    // enforce relayer is whitelisted by calling local connext contract
+    if (!connext.isApprovedRelayer(msg.sender)) revert PromiseRouter__process_notApprovedRelayer();
+
+    // enfore callback fee is not zero
+    if (callbackFees[transferId] == 0) revert PromiseRouter__process_insufficientCallbackFee();
+
+    address callbackAddress = _msg.callbackAddress();
+
+    if (!AddressUpgradeable.isContract(callbackAddress)) revert PromiseRouter__process_notContractCallback();
+
+    (bool success, ) = callbackAddress.call(
+      abi.encodeWithSelector(ICallback.callback.selector, transferId, _msg.returnSuccess(), _msg.returnData())
+    );
+    delete promiseMessages[transferId];
+
+    emit CallbackExecuted(transferId, success, msg.sender);
+
     // Should transfer the stored relayer fee to the msg.sender
+    AddressUpgradeable.sendValue(payable(msg.sender), callbackFees[transferId]);
+    callbackFees[transferId] = 0;
+  }
+
+  /**
+   * @notice This function will be called on the origin domain to increase the callback fee
+   * @param _transferId - The unique identifier of the crosschain transaction
+   */
+  function bumpCallbackFee(bytes32 _transferId) external payable {
```

Comment:
Technically speaking, anyone can bump for transferId that doesn't exist.

While we can typically just say 'ok, but don't bump callback fee for random non-existent transferIds'

there's still a bit of a problem here - there's a race condition where the user may have submitted a bump, but the `process` function has already been called (within the same block, or perhaps 1 block ago, etc.)

So, we should probably check to make sure that this isn't the case.
- We could check to make sure `promiseMessages[transferId]` exists, but that would mean we can only bump fee *after* `handle` has been called, and not prior to handle (e.g. in anticipation that a higher fee may be needed bc market conditions have changed, etc)
- Seeing as how we enforce that the callback fee **must** be > 0 when we call `process`, maybe we can enforce that the callback fee is > 0 when we call send? If so, we could technically check `callbackFee[_transferId] > 0` as a requirement to bump.
- Otherwise, just add a separate mapping for `processedTransferIds` and mark it as true at the end of the `process` call above, and then check it here

Reply:
you should be able to bump the fee before handle is called -- why should you not be able to?

Reply:
ahh okay i see what you're saying -- you can either bump a transfer that doesn't exist or one that was already processed unless there's a flag. 

i dont think we should require the process fees are greater than 0 -- ill go with the flag

Reply:
ahh okay i see what you're saying -- you can either bump a transfer that doesn't exist or one that was already processed unless there's a flag.

i dont think we should require the process fees are greater than 0 -- ill go with the first option as updating 3 mappings seems insane